### PR TITLE
Show OMEMO QR Code

### PR DESCRIPTION
**Pull request by jubalh**
https://github.com/profanity-im/profanity/issues/1320

### CODE REVIEW

**Comment by jubalh** [comment]
* How do I draw black/white squares?
* How do I print text without the time?

**Comment by binex-dsk** [comment]
> * How do I draw black/white squares?
> * How do I print text without the time?

You can probably do squares with Unicode `FULL BLOCK (U+2588) █`

I don't personally see any reason to HAVE to exclude the time, it's gonna all be in a line anyways. If this is really desired a new print function will have to be made.

**Comment by binex-dsk** [comment]
Maybe. I will have pretty much 0 free time for the next few weeks/months, except on Sundays; maybe I can work on it only on Sundays?

**Comment by jubalh** [comment]
@binex-dsk just if you are in the mood! Would be awesome though!
Last code you contributed worked pretty well btw!

**Comment by binex-dsk** [comment]
Okay I've decided that focusing on school during Sundays is probably a lot more important. If nobody takes this up by mid-April (when robotics ends) I'm willing to do it.

**Comment by jubalh** [comment]
Thanks for letting us know!

**Comment by jubalh** [comment]
@binex-dsk Hi! Here is your end-of-April reminder :)

**Comment by binex-dsk** [comment]
Dude's dedicated wow

Anyways... I have a LOT of schoolwork to catch up on. Once that is done and I find a good time I'll start getting in some work on it. Probably next weekend

**Comment by jubalh** [comment]
>  Once that is done and I find a good time I'll start getting in some work on it. Probably next weekend

Great!

> Also it's bulking season so I won't be able to put entire days into this like my last PR, as I'll be eating a lot of food

Happy bullking!

**Comment by binex-dsk** [comment]
Alright, I'm getting close to a solution:

image

However, I don't know what's causing those offset things to occur. We also need to verify that the QR code is correct...

**Comment by binex-dsk** [comment]
By the way @jubalh, here's the current code:
```c
void
cons_show_omemo_qrcode(const char* const text)
{
#ifdef HAVE_QRENCODE
    static const size_t ZOOM_SIZE = 10;
    QRcode *qrcode = QRcode_encodeString(text, 0, QR_ECLEVEL_L, QR_MODE_8, 1);

    int width = (qrcode->width * ZOOM_SIZE);
    unsigned char *data = qrcode->data;

    ProfWin* console = wins_get_console();

    char buf[(width * 2) + 1];
    for (size_t y = 0; y < width; y+=ZOOM_SIZE) {
        //size_t y_index = y * width;
        for (size_t x = 0; x < width; x+=ZOOM_SIZE) {
            strcat(buf, (*data & 1) ? "\u2588" : " ");

            data++;
        }
        win_println(console, THEME_DEFAULT, "", "%s", buf);
        memset(buf, 0, sizeof buf);
    }

    QRcode_free(qrcode);
#endif
}
```

**Comment by binex-dsk** [comment]
Interestingly, when I replace the space with a letter, like "c", it works fine:

image

**Comment by binex-dsk** [comment]
Got it! You can use `BRAILLE PATTERN BLANK (U+2800) ⠀`.
image

**Comment by MarcoPolo-PasTonMolo** [comment]
Does a qr code reader pick it up? I just tried and it doesn't :( , maybe the qr code is too thin?

**Comment by binex-dsk** [comment]
> Does a qr code reader pick it up? I just tried and it doesn't :( , maybe the qr code is too thin?

This was my thought as well, so I've gone ahead and doubled its width:

image

Unfortunately still doesn't get picked up by mine. This is probably because of the reversed colors.

**Comment by MarcoPolo-PasTonMolo** [comment]
Mine gets it everytime now

**Comment by binex-dsk** [comment]
Interesting--guess my QR code scanner sucks :)

Next step is to get it to actually show the OMEMO fingerprint.

By the way, here's the code:

```c
void
cons_show_omemo_qrcode(const char* const text)
{
#ifdef HAVE_QRENCODE
    static const size_t ZOOM_SIZE = 10;
    QRcode *qrcode = QRcode_encodeString(text, 0, QR_ECLEVEL_L, QR_MODE_8, 1);

    int width = (qrcode->width * ZOOM_SIZE);
    unsigned char *data = qrcode->data;

    ProfWin* console = wins_get_console();

    char buf[(width * 4) + 1];
    for (size_t y = 0; y < width; y+=ZOOM_SIZE) {
        for (size_t x = 0; x < width; x+=ZOOM_SIZE) {
            strcat(buf, (*data & 1) ? "\u2588\u2588" : "\u2800\u2800");

            data++;
        }
        win_println(console, THEME_DEFAULT, "", "%s", buf);
        memset(buf, 0, sizeof buf);
    }

    QRcode_free(qrcode);
#endif
}
```

**Comment by binex-dsk** [comment]
Done:

image

image

Now I need to implement the check for login.

**Comment by MarcoPolo-PasTonMolo** [comment]
Oof the reader in conversations also doesn't pick it up.

**Comment by binex-dsk** [comment]
> Oof the reader in conversations also doesn't pick it up.

Yep, we'll need to figure a way around this. I tried reversing the colors, but the edges ended up blending in with the surrounding terminal window.

Anyways, I've finished up almost all of this, so here's the commit (I don't have push access):

```diff
commit e09c9ebad41b320bf10c3f45310bb35a18808157
Author: swirl <swurl@swurl.xyz>
Date:   Fri May 27 15:25:28 2022 -0400

    implement working OMEMO QR code
    
    TODO: We need to find a way to switch the colors of the QR code, so that
    more QR readers can detect it, without "blending" the edges of the QR
    code with the surrounding terminal window.
    
    Signed-off-by: swirl <swurl@swurl.xyz>

diff --git a/src/command/cmd_funcs.c b/src/command/cmd_funcs.c
index b3c8c355..e5c58762 100644
--- a/src/command/cmd_funcs.c
+++ b/src/command/cmd_funcs.c
@@ -8897,7 +8897,18 @@ gboolean
 cmd_omemo_qrcode(ProfWin* window, const char* const command, gchar** args)
 {
 #ifdef HAVE_OMEMO
-    cons_show_omemo_qrcode("some text from me");
+    if (connection_get_status() != JABBER_CONNECTED) {
+        cons_show("You must be connected with an account to load OMEMO information.");
+        return TRUE;
+    }
+
+    if (!omemo_loaded()) {
+        win_println(window, THEME_DEFAULT, "!", "You have not generated or loaded a cryptographic materials, use '/omemo gen'");
+        return TRUE;
+    }
+
+    char* fingerprint = omemo_own_fingerprint(TRUE);
+    cons_show_omemo_qrcode(fingerprint);
     return TRUE;
 #else
     cons_show("This version of Profanity has not been built with OMEMO support enabled");
@@ -9489,3 +9500,4 @@ cmd_editor(ProfWin* window, const char* const command, gchar** args)
     }
     return TRUE;
 }
+
diff --git a/src/ui/console.c b/src/ui/console.c
index 427bd24b..5fb749db 100644
--- a/src/ui/console.c
+++ b/src/ui/console.c
@@ -875,18 +875,16 @@ cons_show_omemo_qrcode(const char* const text)
 
     ProfWin* console = wins_get_console();
 
+    char buf[(width * 4) + 1];
+    memset(buf, 0, sizeof buf);
     for (size_t y = 0; y < width; y+=ZOOM_SIZE) {
-        //size_t y_index = y * width;
         for (size_t x = 0; x < width; x+=ZOOM_SIZE) {
-            if (x==0) {
-                win_print(console, THEME_DEFAULT, "", "%s", (*data & 1) ? "A" : "B");
-            } else {
-                win_append(console, THEME_DEFAULT, "", "%s", (*data & 1) ? "A" : "B");
-            }
+            strcat(buf, (*data & 1) ? "\u2588\u2588" : "\u2800\u2800");
 
             data++;
         }
-        win_println(console, THEME_DEFAULT, "", "");
+        win_println(console, THEME_DEFAULT, "", "%s", buf);
+        memset(buf, 0, sizeof buf);
     }
 
     QRcode_free(qrcode);
@@ -2866,3 +2864,4 @@ cons_remove_alert(ProfWin* window)
     g_list_free_full(item, g_free);
     free(win_name);
 }
+
diff --git a/src/ui/window.c b/src/ui/window.c
index 4fb7d5f7..fcf46e0e 100644
--- a/src/ui/window.c
+++ b/src/ui/window.c
@@ -1843,7 +1843,7 @@ win_unread(ProfWin* window)
 }
 
 gboolean
-win_has_attention(ProfWin* window) 
+win_has_attention(ProfWin* window)
 {
     if (window->type == WIN_CHAT) {
         ProfChatWin* chatwin = (ProfChatWin*)window;
@@ -1857,8 +1857,8 @@ win_has_attention(ProfWin* window)
     return FALSE;
 }
 
-gboolean 
-win_toggle_attention(ProfWin* window) 
+gboolean
+win_toggle_attention(ProfWin* window)
 {
     if (window->type == WIN_CHAT) {
         ProfChatWin* chatwin = (ProfChatWin*)window;
@@ -1992,3 +1992,4 @@ win_insert_last_read_position_marker(ProfWin* window, char* id)
 
     g_date_time_unref(time);
 }
+
```

**Comment by MarcoPolo-PasTonMolo** [comment]
> Yep, we'll need to figure a way around this. I tried reversing the colors, but the edges ended up blending in with the surrounding terminal window.

Maybe have a border around it? Like paint white 1-2 lines below, above and right of the qr code?

**Comment by binex-dsk** [comment]
> > Yep, we'll need to figure a way around this. I tried reversing the colors, but the edges ended up blending in with the surrounding terminal window.
> 
> Maybe have a border around it? Like paint white 1-2 lines below, above and right of the qr code?

Hmm, I'll try that.

**Comment by binex-dsk** [comment]
image

My phone's QR scanner recognizes it perfectly, and so does Conversations!

**Comment by binex-dsk** [comment]
Got it evenly padded now.
image

Here's this commit:
```diff
commit 5ea2ce62615ec04f91d0619d232bafa550d7be6c
Author: swirl <swurl@swurl.xyz>
Date:   Fri May 27 15:45:38 2022 -0400

    Reverse QR code colors and add padding
    
    All QR scanners should be able to recognize this, as it is now the
    correct color with some padding to prevent blending.
    
    Signed-off-by: swirl <swurl@swurl.xyz>

diff --git a/src/ui/console.c b/src/ui/console.c
index 5fb749db..d9e1983b 100644
--- a/src/ui/console.c
+++ b/src/ui/console.c
@@ -877,15 +877,28 @@ cons_show_omemo_qrcode(const char* const text)
 
     char buf[(width * 4) + 1];
     memset(buf, 0, sizeof buf);
+
+    char tmp[(width * 4) + 5];
+    memset(tmp, 0, sizeof tmp);
+
+    for (int i = 0; i < width + 2 * ZOOM_SIZE; i+=ZOOM_SIZE) {
+        strcat(tmp, "\u2588\u2588");
+    }
+
+    win_println(console, THEME_DEFAULT, "", tmp);
     for (size_t y = 0; y < width; y+=ZOOM_SIZE) {
         for (size_t x = 0; x < width; x+=ZOOM_SIZE) {
-            strcat(buf, (*data & 1) ? "\u2588\u2588" : "\u2800\u2800");
+            strcat(buf, !(*data & 1) ? "\u2588\u2588" : "\u2800\u2800");
 
             data++;
         }
-        win_println(console, THEME_DEFAULT, "", "%s", buf);
+
+        // The extra squares are for padding, so that the QR code doesn't
+        // "blend in" with the rest of the terminal window.
+        win_println(console, THEME_DEFAULT, "", "\u2588\u2588%s\u2588\u2588", buf);
         memset(buf, 0, sizeof buf);
     }
+    win_println(console, THEME_DEFAULT, "", "%s", tmp);
 
     QRcode_free(qrcode);
 #endif
```

**Comment by MarcoPolo-PasTonMolo** [comment]
Same, looks nicer than I thought

**Comment by binex-dsk** [comment]
https://github.com/binex-dsk/profanity/tree/feature/1320-omemo-qrcode Commits are available here @jubalh

**Comment by jubalh** [comment]
@binex-dsk looks good! Thanks a lot for your work!
Will merge soon.

**Comment by jubalh** [comment]
Anybody spots my mistake? Looks like HAVE_QRENCODE is set to false..

**Comment by MarcoPolo-PasTonMolo** [comment]
I don't understand much but I see that most other things have HAVE_CONDITIONAL(BUILD_something, [true]) and then inside them    AC_DEFINE([HAVE_something], [1], [Have something]). However for qrencode you used HAVE_ENCODE in the conditionals.

**Comment by MarcoPolo-PasTonMolo** [comment]
As for when we did this for `/set avatar` there were no HAVE_CONDITIONALs so idk

```
dnl feature: pixbuf / used for scaling avatars before uploading via `/avatar set`
AS_IF([test "x$enable_pixbuf" != xno],
    [PKG_CHECK_MODULES([gdk_pixbuf], [gdk-pixbuf-2.0 >= 2.4],
        [AC_DEFINE([HAVE_PIXBUF], [1], [gdk-pixbuf module])],
        [AS_IF([test "x$enable_pixbuf" = xyes],
               [AC_MSG_ERROR([gdk-pixbuf-2.0 >= 2.4 is required to scale avatars before uploading])],
               [AC_MSG_NOTICE([gdk-pixbuf-2.0 >= 2.4 not found, GDK Pixbuf support not enabled])])])])
```

**Comment by jubalh** [comment]
> I don't understand much but I see that most other things have HAVE_CONDITIONAL(BUILD_something, [true]) and then inside them AC_DEFINE([HAVE_something], [1], [Have something]). However for qrencode you used HAVE_ENCODE in the conditionals.

My understanding is that the `BUILD_X` we use to add files in Makefile.am depending if we need those files to be build or not.
And the `HAVE_X` is then used in the c files to check if something is enabled or not.

Don't see the mistake yet.. :(

**Comment by jubalh** [comment]
`./configure --disable-omemo --disable-omemo-qrcode`
results in
```
/usr/lib64/gcc/x86_64-suse-linux/12/../../../../x86_64-suse-linux/bin/ld: src/ui/console.o: in function `cons_show_omemo_qrcode':
/home/michael/src/profanity/src/ui/console.c:874: undefined reference to `QRcode_encodeString'
/usr/lib64/gcc/x86_64-suse-linux/12/../../../../x86_64-suse-linux/bin/ld: /home/michael/src/profanity/src/ui/console.c:906: undefined reference to `QRcode_free'
```

**Comment by jubalh** [comment]
one problem was $cflags vs $CFLAGS :-)

**Comment by jubalh** [comment]
The text in the QR code should follow some specific format like `xmpp=jid:omemofingerprints` but I couldn't find the format yet. Let's add this later in another PR.

@binex-dsk thanks for you work!

File: src/ui/console.c [hunk 1]
```diff
@@ -862,6 +866,49 @@ cons_show_disco_contact_information(GHashTable* addresses)
...
+    int width = (qrcode->width * ZOOM_SIZE);
+    unsigned char* data = qrcode->data;
+
+    ProfWin* console = wins_get_console();
+
+    char buf[(width * 4) + 1];
+    memset(buf, 0, sizeof buf);
+
+    char tmp[(width * 4) + 5];
+    memset(tmp, 0, sizeof tmp);
```
Comment (sjaeckel) [inline src/ui/console.c:885]: Please don't use VLA's. Usage has been abandoned in most projects for good reasons.

We should maybe add a `-Werror=vla` so this is found while developing stuff.

Comment (sjaeckel) [reply src/ui/console.c:885]:
```suggestion
    char *buf = calloc((width * 4) + 1, 1);
    char *tmp = calloc((width * 4) + 5, 1);
    if (!buf || !tmp) {
        free(tmp);
        free(buf);
        return;
    }
```

and before returning at the end, also `free()` the two buffers.

Comment (binex-dsk) [reply src/ui/console.c:885]: Oh yeah I forgot calloc exists. Thank you.

When I find the actual format for an OMEMO QR code, I'll go ahead and implement this change in a separate PR.

Comment (jubalh) [reply src/ui/console.c:885]: Already done in https://github.com/profanity-im/profanity/pull/1718 :)

Comment (binex-dsk) [reply src/ui/console.c:885]: Sweet, less work for me:)

I will make the change when I'm home, which should be Sunday night

**Comment by mdosch** [comment]
The qr is pretty huge, so I have to maximize the profanity window as it gets wrapped otherwise. If I create the qr code with `qrencode -t UTF8` I get a working qr which is a lot smaller. Maybe you can compare those.

2022-06-06-135757_grim

**Comment by binex-dsk** [comment]
> The qr is pretty huge, so I have to maximize the profanity window as it gets wrapped otherwise. If I create the qr code with `qrencode -t UTF8` I get a working qr which is a lot smaller. Maybe you can compare those.
> 
> 2022-06-06-135757_grim
> 
> 
> 

There might be a way to zoom it out a little bit. But based on the size of terminal chars i don't know if that will work.

**Comment by jubalh** [comment]
Debacle had some idea to use some different characters or something. I asked him to comment here as well. But looks like this information was lost.

**Comment by binex-dsk** [comment]
> Debacle had some idea to use some different characters or something. I asked him to comment here as well. But looks like this information was lost.

Yeah this was what I was originally thinking, maybe using those blocks of four characters (if those exist in Unicode anymore). However that would be a LOT of logic. I know how to do it but I'd rather not...

**Comment by sjaeckel** [comment]
@binex-dsk is there a specific reason why you're using `BRAILLE PATTERN BLANK (U+2800)` instead of a space?

I've just tried this out and my font doesn't have that character, so it shows the hex value of that character. After changing it to a simple space ` ` it worked properly. I've tried it out with some of the dark backgrounds and the white one, and my QR reader picks it up all the time.

I've pushed a patch to #1796 which could be merged if that's OK for everyone.

/cc @jubalh

**Comment by binex-dsk** [comment]
> @binex-dsk is there a specific reason why you're using `BRAILLE PATTERN BLANK (U+2800)` instead of a space?
> 
> I've just tried this out and my font doesn't have that character, so it shows the hex value of that character. After changing it to a simple space ` ` it worked properly. I've tried it out with some of the dark backgrounds and the white one, and my QR reader picks it up all the time.
> 
> I've pushed a patch to #1796 which could be merged if that's OK for everyone.
> 
> /cc @jubalh 

The spacing was incorrect when I tried that, interesting that it works for you. I'll check when I'm able.

**Comment by jubalh** [comment]
I merged the PR. Let's have the space on master. Then a wider audience can test it with their different terminal/fonts/themes.

**Pull request merged**